File: docstrings/bbox.py
# -*- coding: utf-8 -*-

# Copyright (c) 2015, Michael Droettboom All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:

# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# The views and conclusions contained in the software and
# documentation are those of the authors and should not be interpreted
# as representing official policies, either expressed or implied, of
# the FreeBSD Project.

from __future__ import print_function, unicode_literals, absolute_import

BBox__init__ = """
An outline's bounding box.

The bounding box is specified with the coordinates of the lower left
and the upper right corner. In PostScript, those values are often
called ``(llx, lly)`` and ``(urx, ury)``, respectively.

If `y_min` is negative, this value gives the glyph's
descender. Otherwise, the glyph doesn't descend below the
baseline. Similarly, if `y_max` is positive, this value gives the
glyph's ascender.

`x_min` gives the horizontal distance from the glyph's origin to the
left edge of the glyph's bounding box. If `x_min` is negative, the
glyph extends to the left of the origin.

`BBox` also works as a Python sequence, so it is easy to do::

    x_min, y_min, x_max, y_max = bbox
"""

BBox_ascent = """
The height of the bounding box above the baseline.  This is an alias
for `y_max`.
"""

BBox_depth = """
The depth of the bounding box below the baseline.  This is an alias
for `y_min`.
"""

BBox_height = """
The height of the bounding box (``y_max - y_min``).
"""

BBox_width = """
The width of the bounding box (``x_max - x_min``).
"""

BBox_x_max = """
The horizontal maximum (right-most).
"""

BBox_x_min = """
The horizontal minimum (left-most).
"""

BBox_y_max = """
The vertical maximum (top-most).
"""

BBox_y_min = """
The vertical minimum (bottom-most).
"""

File: src/bbox.c
/*
Copyright (c) 2013, Michael Droettboom
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of the FreeBSD Project.
*/

#include "bbox.h"
#include "doc/bbox.h"

#define BBOX_GETTER(name, type) GETTER(name, type, FT_BBoxRec, Py_BBox);
#define DEF_BBOX_GETTER(name) DEF_GETTER(name, doc_BBox_ ## name)


/****************************************************************************
 Object basics
*/


typedef struct {
    ftpy_Object base;
    double xMin;
    double xMax;
    double yMin;
    double yMax;
} Py_BBox;


static PyTypeObject Py_BBox_Type;


PyObject *
Py_BBox_cnew(FT_BBox *bbox, double scale)
{
    Py_BBox *self;
    self = (Py_BBox *)(&Py_BBox_Type)->tp_alloc(&Py_BBox_Type, 0);
    if (self == NULL) {
        return NULL;
    }
    self->base.owner = NULL;
    self->xMin = bbox->xMin * scale;
    self->xMax = bbox->xMax * scale;
    self->yMin = bbox->yMin * scale;
    self->yMax = bbox->yMax * scale;
    return (PyObject *)self;
}


static PyObject *
Py_BBox_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    Py_BBox *self;

    self = (Py_BBox *)ftpy_Object_new(type, args, kwds);
    if (self == NULL) {
        return NULL;
    }
    self->xMin = 0;
    self->xMax = 0;
    self->yMin = 0;
    self->yMax = 0;
    return (PyObject *)self;
}


static int
Py_BBox_init(Py_BBox *self, PyObject *args, PyObject *kwds)
{
    PyErr_SetString(
        PyExc_RuntimeError,
        "BBox objects may not be instantiated directly. "
        "Use Face.bbox instead.");
    return -1;
}


/****************************************************************************
 Get/Set
*/


static PyObject *depth_get(Py_BBox *self, PyObject *closure)
{
    return PyFloat_FromDouble(self->yMin);
}


static PyObject *height_get(Py_BBox *self, PyObject *closure)
{
    return PyFloat_FromDouble(self->yMax);
}


static PyObject *width_get(Py_BBox *self, PyObject *closure)
{
    return PyFloat_FromDouble(self->xMax - self->xMin);
}


static PyGetSetDef Py_BBox_getset[] = {
    DEF_BBOX_GETTER(depth),
    DEF_BBOX_GETTER(height),
    DEF_BBOX_GETTER(width),
};


/****************************************************************************
 Members
*/


static PyMemberDef Py_BBox_members[] = {
    {"x_min", T_DOUBLE, offsetof(Py_BBox, xMin), READONLY, doc_BBox_x_min},
    {"x_max", T_DOUBLE, offsetof(Py_BBox, xMax), READONLY, doc_BBox_x_max},
    {"y_min", T_DOUBLE, offsetof(Py_BBox, yMin), READONLY, doc_BBox_y_min},
    {"y_max", T_DOUBLE, offsetof(Py_BBox, yMax), READONLY, doc_BBox_y_max},
    {NULL}  /* Sentinel */
};


/****************************************************************************
 Sequence
*/


static Py_ssize_t Py_BBox_len(Py_BBox *self)
{
    return 4;
}


static PyObject *Py_BBox_getitem(Py_BBox *self, Py_ssize_t i)
{
    if (i == 0) {
        return PyFloat_FromDouble(self->xMin);
    } else if (i == 1) {
        return PyFloat_FromDouble(self->yMin);
    } else if (i == 2) {
        return PyFloat_FromDouble(self->xMax);
    } else if (i == 3) {
        return PyFloat_FromDouble(self->yMax);
    }

    PyErr_SetString(PyExc_IndexError, "index out of range");
    return NULL;
}


static PySequenceMethods Py_BBox_sequence;


/****************************************************************************
 repr
*/


static PyObject *Py_BBox_repr(PyObject *self)
{
    PyObject *tuple = NULL;
    PyObject *result = NULL;

    tuple = PySequence_Tuple(self);
    if (tuple == NULL) {
        goto exit;
    }

    result = PyObject_Repr(tuple);

 exit:
    Py_XDECREF(tuple);

    return result;
}


/****************************************************************************
 Setup
*/


static PyTypeObject Py_BBox_Type;


int setup_BBox(PyObject *m)
{
    memset(&Py_BBox_sequence, 0, sizeof(PySequenceMethods));
    Py_BBox_sequence = (PySequenceMethods) {
        .sq_length = (lenfunc)Py_BBox_len,
        .sq_item = (ssizeargfunc)Py_BBox_getitem
    };

    memset(&Py_BBox_Type, 0, sizeof(PyTypeObject));
    Py_BBox_Type = (PyTypeObject) {
        .tp_name = "freetypy.BBox",
        .tp_basicsize = sizeof(Py_BBox),
        .tp_as_sequence = &Py_BBox_sequence,
        .tp_repr = Py_BBox_repr,
        .tp_doc = doc_BBox__init__,
        .tp_members = Py_BBox_members,
        .tp_getset = Py_BBox_getset,
        .tp_init = (initproc)Py_BBox_init,
        .tp_new = Py_BBox_new
    };

    if (ftpy_setup_type(m, &Py_BBox_Type)) {
        return -1;
    }

    return 0;
}
